### DISCUSSION TITLE
The _query attribute

### DISCUSSION THREAD

**Discussion by LongBeachHXC** [psycopg/psycopg#88] [General]
I was leveraging the `cur.query` and `cur.params` attribute as a sanity check. I was doing some digging and noticed the changes to the Cursor API. I'm curious why NOT provide this information? I would love to be able to leverage this information as a sanity check to make sure the correct statement is being generated and being sent down the pipe.

I may be missing something obvious here but I couldn't find the equivalent as it is. I shifted to use the `cur.statusmessage` but it doesn't give me the level of granularity that I would like to see.

**Comment by dvarrazzo** [comment]
`query` and `params` are incomplete information: you would also need at least `types`, `formats` to reconstruct what's really passed to the server. So we expose the whole `PostgresQuery` object responsible of the transformation of the Python query/params into Postgres format, instead of bloating the Cursor interface adding all the details.

However the query object is still an internal object and I prefer to not commit altogether to its interface: if there are valid reasons it might change in the future. Possible reasons might be 1) performance and 2) make it an abstract interface and provide different implementations, such as a client-side binding query too, to give an option to users finding server-side binding problematic for their use cases.

For this reason the object is only exposed using a `_query` private-ish attribute. The object is still accessible for debug and exploration reasons (and care has been taken in order to make it available when needed, e.g. including on error and after the cursor is closed), but I'd rather discourage people from building entire services on top of it, before maybe having a chat and designing a more stable interface.

Does it make sense? What is your use case?

**Reply by jankatins** [reply]
One thing where it would be nice to expose the query would be a syntax error: PG shows you something (the line which is wrong), but if the query is generated on the fly, having the full query would be a lot nicer...

E.g. I have this error, but my query actually looks ok:

```
E           psycopg.errors.SyntaxError: syntax error at or near "order"
E           LINE 8:             order by task.task_id asc
```

My query looked like this:

```py
            select
                ...
            from task
            where ...
              and {pagination_where_condition}
            order by task.task_id asc
            fetch first %(number_of_rows)s rows only;
```

In the end it turned out that I did the brackets in the pagination_where_condition wrong... Easy enough after going through the code , but with the query, it would have been even easier...

**Reply by dvarrazzo** [reply]
@jankatins you can catch the exception in the context where it happens and re-raise with the full query. Doing it for everyone might not be what people want. The query might contain secrets, or being too large, or they may want it in their log system instead of on stderr, or the other way round...

You are in control of the place where the exception happens and you have access to the cursor, so you can print the query for debugging precisely the way you want.

**Comment by bluetech** [comment]
I've opened a Django issue that's related to this discussion: https://code.djangoproject.com/ticket/35021
